Convert Find Coronavirus Support flow tests

leenagupte opened the pull request:
Trello: https://trello.com/c/6h5fSoof

# What's changed and why?

Converts the Find Coronavirus Support flow tests to use the
node based testing approach introduced in [ADR-3][1].

Some of the links are also being explicitly tested in the outcome as
much of the conditional guidance, especially for England and Wales
has the same link text.

Also removes the existing RSpec feature tests as they are no longer
needed.

[1]: https://github.com/alphagov/smart-answers/blob/main/docs/arch/003-testing-flows.md

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

JonathanHallam commented on test/flows/find_coronavirus_support_flow_test.rb:
```diff
@@ -0,0 +1,597 @@
+      end
+
+      should "have next node of next question group for #{response} if there are other needs_help_with questions" do
+        add_responses need_help_with: %w[paying_bills mental_health]
```
Is the `paying_bills` here not redundant because it's added on line 57?

JonathanHallam replied:
I deleted it and re-ran the tests and a load fail so short answer, no it's not redundant

JonathanHallam commented on test/flows/find_coronavirus_support_flow_test.rb:
```diff
@@ -0,0 +1,597 @@
+        assert_next_node :nation, for_response: response
+      end
+
+      should "have next node of next question group for #{response} if there are other needs_help_with questions" do
```
I think we're okay to change "next question group" to "mental health worries" given that it's hardcoded below. Figure it makes the test titles a bit clearer. Applies to all the tests so I'm not going to repeat this over and over

JonathanHallam replied:
ah, ignore this comment, I've realised why i'm wrong 🤦

kevindew commented on test/flows/find_coronavirus_support_flow_test.rb:
```diff
@@ -0,0 +1,597 @@
+      setup do
+        add_responses need_help_with: %w[somewhere_to_live],
+                      have_somewhere_to_live: "no",
+                      have_you_been_evicted: "no",
```
As per above about that this should also default to yes, given there is a ton of conditionals within it (no need to test them, just nice that the code executes)

kevindew reviewed (commented):
Awesome job Leena 👍  Looks great

I've got a few comments that I've added only once but will apply in multiple places. Happy to chat through any.

JonathanHallam commented on test/flows/find_coronavirus_support_flow_test.rb:
```diff
@@ -0,0 +1,597 @@
+    end
+  end
+
+  context "question: self_employed" do
```
Do you think it'd be worthwhile to add an extra context block for `being unemployed` that wraps around this? Figure it would give a bit more context around how the code is structured which might be useful if any new folk come into this codebase. Also applies to the other groups

leenagupte replied:
Sounds good to me. @kevindew How does this suggestion fit with the convention of the new test structure?

kevindew replied:
It's unfortunately counter to the approach of structuring these flow tests and I don't think you should do it.

I can see the logic of the suggestion, these questions are both related to a journey through this of unemployment, however I think that is just the tip of the iceberg for Smart Answers as many nodes are only related to journeys of other ones. I think the only consistent logical approach is that given nodes are a top level concept then they should only be tested as top level concepts.

I don't see the benefit of that myself. Both self_employed and have_you_been_made_unemployed are top level nodes so I don't see why they should be nested. 

That seems like a slippery slope towards context based on user journeys that can easily multiply into the thousands.

leenagupte commented on test/flows/find_coronavirus_support_flow_test.rb:
```diff
@@ -0,0 +1,597 @@
+    end
+
+    %w[yes concerned_about_others no not_sure].each do |response|
+      should "have a next node of nation if response is #{response} and there aren't any other needs_help_with questions" do
```
Oops! Will add them in.

leenagupte commented on test/flows/find_coronavirus_support_flow_test.rb:
```diff
@@ -0,0 +1,597 @@
+      assert_rendered_question
+    end
+
+    %w[yes concerned_about_others no not_sure].each do |response|
```
I'm not sure what the convention is. Both the `afford_food` question and `feel_unsafe` are part of a conditional in the calculator. So, should I leave this loop in, or take them both out?

kevindew replied:
Sure, the convention is to exercise the logic of the flow. So if the calculator has logic but the flow doesn't the flow tests don't need to test that logic as that should be covered by calculator unit tests.

I'd take all the loops out that don't involve any logic in the flow. The only question that needs more than one test for the next node in this is self employed since it has a conditional: https://github.com/alphagov/smart-answers/blob/396f93306c6ad512925851557265a22ae79042e1/app/flows/find_coronavirus_support_flow.rb#L116-L120

leenagupte commented on test/flows/find_coronavirus_support_flow_test.rb:
```diff
@@ -0,0 +1,597 @@
+    context "being_unemployed" do
+      setup do
+        add_responses need_help_with: %w[being_unemployed],
+                      self_employed: "no",
```
This won't work. If self_employed is "yes", the have_you_been_made_unemployed question is skipped:

next_node do
        if calculator.self_employed == "yes"
          question calculator.next_question(:have_you_been_made_unemployed)
        else
          question :have_you_been_made_unemployed
        end
      end


So to test rendering of the have_you_been_made_unemployed results, self_employed needs to be "no".

I did it in this order, i.e. have_you_been_made_unemployed first as have_you_been_made_unemployed is the first calculator value checked in the being_unemployed partial.

Also, I thought it was ok to only test the nation specific content once.

leenagupte replied:
I suppose I could put the tests in separate sub contexts? One where `self_employed` is "yes" and the other where it's "no" to test the `have_you_been_made_unemployed` outcome. I think that might make it easier to read, though I'm not sure I relish the idea of having two sets of nation checks.

kevindew replied:
Oh I see, thanks for explaining. I hadn't realised that it was actually a one or the other for unemployed or self employed rather than that both can be available (shame the code wasn't an `else`). I guess with hindsight this should probably be two different partials.

> Also, I thought it was ok to only test the nation specific content once.

Yeah, that seemed a reasonable approach if testing one meant they were all executed. We've got that slightly annoying situation here where, by only testing them once for this partial, that either the unemployed conditionals are executed or the self employed ones are. So we lose some coverage either way.

I can understand not wanting to do all of the nation conditionals again and this seems sufficiently nuanced that it's not something that could be enforced into future or explained. So don't worry about them if we can't express a way that they add value.

leenagupte replied:
So, do you think it would be better to split the `being_unemployed` tests into separate contexts, or are they ok as they are?

kevindew replied:
If it's not too much of a pain I think it'd be better to split the being_unemployed into a separate context. I think it would be good to satisfy the convention of full coverage through these files.

leenagupte commented:
@kevindew, @JonathanHallam Thank you for your reviews. I've updated and fixed up most of the suggestions, and left follow-up questions about the others.

Also, I forgot to delete the integration tests 😱, so I've done that as well.

leenagupte commented on test/flows/find_coronavirus_support_flow_test.rb:
```diff
@@ -0,0 +1,597 @@
+      assert_rendered_question
+    end
+
+    %w[yes no not_sure].each do |response|
```
So this loop should be taken out as well.

leenagupte commented:
@kevindew I've removed the extra loops from the tests and fixed up.

leenagupte commented:
@kevindew thanks for your comments. I've split up the `being_unemployed` tests and fixed up.

kevindew reviewed (approved):
Nice one
